File: src/main/java/com/vc/medium/L347.scala
package com.vc.medium

object L347 {
  object PQOrdering extends Ordering[(Int, Int)] {
    def compare(x:(Int, Int), y:(Int, Int)): Int = x._2.compareTo(y._2)
  }
  def topKFrequent(arr: Array[Int], k: Int): List[Int] = {
    import scala.collection.mutable
    val map = new mutable.HashMap[Int, Int]()
    val pq = new mutable.PriorityQueue[(Int, Int)]()(PQOrdering)
    arr.foreach(x => map.put(x, map.getOrElse(x, 0) + 1))
    map.foreach(x => pq += x)
    val list = new mutable.ListBuffer[Int]()
    var i = 0
    while(i < k) {
      list += pq.dequeue._1
      i += 1
    }
    list.toList
  }
}

File: src/main/java/com/vc/medium/L347.java
package com.vc.medium;

import java.util.*;

class L347 {
    public List<Integer> topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<int[]>(new Comparator<int[]>(){
            public int compare(int[] x, int[] y) {
                return Integer.compare(y[0], x[0]);
            }
        });

        for(Map.Entry<Integer, Integer> entry: map.entrySet()) {
            pq.offer(new int[]{entry.getValue(), entry.getKey()});
        }

        List<Integer> res = new ArrayList<>();
        while(k-- > 0) res.add(pq.poll()[1]);
        return res;
    }
}
